elrond-wasm/src/storage/mappers/token_attributes_mapper.rs

```diff
@@ -0,0 +1,239 @@
+            self.set_counter_value(counter);
+            counter
+        };
+
+        let has_value = !self.is_empty_token_attributes_value(mapping, token_nonce);
+        if has_value {
+            self.api.signal_error(VALUE_ALREADY_SET_ERROR_MESSAGE);
+        }
+
+        let mut value = Vec::<u8>::new();
+        attributes.top_encode(&mut value).unwrap();
+        self.set_token_attributes_value(mapping, token_nonce, value);
+    }
+
+    ///Use carefully. Update should be used mainly when backed up by the protocol.
+    pub fn update<T: elrond_codec::TopEncode + elrond_codec::TopDecode>(
+        &self,
+        token_id: &TokenIdentifier,
+        token_nonce: u64,
+        attributes: &T,
+    ) {
+        let has_mapping = !self.is_empty_mapping_value(token_id);
+        if !has_mapping {
+            self.api.signal_error(UNKNOWN_TOKEN_ID_ERROR_MESSAGE);
+        }
+
+        let mapping = self.get_mapping_value(token_id);
+        let has_value = !self.is_empty_token_attributes_value(mapping, token_nonce);
+        if !has_value {
+            self.api
+                .signal_error(VALUE_NOT_PREVIOUSLY_SET_ERROR_MESSAGE);
+        }
+
+        let mut value = Vec::<u8>::new();
+        attributes.top_encode(&mut value).unwrap();
+        self.set_token_attributes_value(mapping, token_nonce, value)
+    }
+
+    pub fn clear(&self, token_id: &TokenIdentifier, token_nonce: u64) {
+        let has_mapping = !self.is_empty_mapping_value(token_id);
+        if !has_mapping {
+            self.api.signal_error(UNKNOWN_TOKEN_ID_ERROR_MESSAGE);
+        }
+
+        let mapping = self.get_mapping_value(token_id);
+        let has_value = !self.is_empty_token_attributes_value(mapping, token_nonce);
+        if !has_value {
+            self.api
+                .signal_error(VALUE_NOT_PREVIOUSLY_SET_ERROR_MESSAGE);
+        }
+
+        self.clear_token_attributes_value(mapping, token_nonce);
+    }
+
+    pub fn is_empty(&self, token_id: &TokenIdentifier, token_nonce: u64) -> bool {
+        let has_mapping = self.is_empty_mapping_value(token_id);
+        let mapping = self.get_mapping_value(token_id);
+
+        !has_mapping || self.is_empty_token_attributes_value(mapping, token_nonce)
+    }
+
+    pub fn get<T: elrond_codec::TopEncode + elrond_codec::TopDecode>(
+        &self,
+        token_id: &TokenIdentifier,
+        token_nonce: u64,
+    ) -> T {
+        let has_mapping = !self.is_empty_mapping_value(token_id);
+        if !has_mapping {
+            self.api.signal_error(UNKNOWN_TOKEN_ID_ERROR_MESSAGE);
+        }
+
+        let mapping = self.get_mapping_value(token_id);
+        let has_value = !self.is_empty_token_attributes_value(mapping, token_nonce);
+        if !has_value {
+            self.api
+                .signal_error(VALUE_NOT_PREVIOUSLY_SET_ERROR_MESSAGE);
+        }
+
+        let value = self.get_token_attributes_value(mapping, token_nonce);
+        T::top_decode(value).unwrap()
```

Comment:
Don't manually decode and avoid `unwrap` (since it panics).
Use the power of generics to extract the right type directly from `storage_get`.
`storage_get` will also handle errors for you, plus it knows how to handle managed types, which `Vec<u8>` does not.